src/backtrace/unwind.rs

```diff
@@ -20,60 +20,67 @@ static MISSING_DEBUG_INFO: &str = "debug information is missing. Likely fixes:
 
 /// Virtually* unwinds the target's program
 /// \* destructors are not run
-// FIXME(?) this should be "infallible" and return as many frames as possible even in case of IO
-// errors
+///
+/// This returns as much info as could be collected, even if the collection is interrupted by an error.
+/// if an error occurred during processing, it is stored in `Output::processing_error`.
 pub(crate) fn target(
     core: &mut Core,
     debug_frame: &[u8],
     vector_table: &VectorTable,
     sp_ram_region: &Option<RamRegion>,
-) -> anyhow::Result<Output> {
+) -> Output {
+    let mut output = Output {
+        corrupted: true,
+        outcome: Outcome::Ok,
+        raw_frames: vec![],
+        processing_error: None,
+    };
+
+    // returns all info collected until the error occurred and puts the error into `processing_error`
+    macro_rules! unwrap_or_return_output {
+        ( $e:expr ) => {
+            match $e {
+                Ok(x) => x,
+                Err(err) => {
+                    output.processing_error = Some(anyhow!(err));
+                    return output;
+                }
+            }
+        };
+    }
+
     let mut debug_frame = DebugFrame::new(debug_frame, LittleEndian);
     debug_frame.set_address_size(cortexm::ADDRESS_SIZE);
 
-    let mut pc = core.read_core_reg(registers::PC)?;
-    let sp = core.read_core_reg(registers::SP)?;
-    let lr = core.read_core_reg(registers::LR)?;
+    let mut pc = unwrap_or_return_output!(core.read_core_reg(registers::PC));
+    let sp = unwrap_or_return_output!(core.read_core_reg(registers::SP));
+    let lr = unwrap_or_return_output!(core.read_core_reg(registers::LR));
     let base_addresses = BaseAddresses::default();
     let mut unwind_context = UninitializedUnwindContext::new();
-
-    let mut outcome = Outcome::Ok;
     let mut registers = Registers::new(lr, sp, core);
-    let mut raw_frames = vec![];
-    let mut corrupted = true;
 
     loop {
-        if cortexm::is_hard_fault(pc, vector_table) {
-            assert!(
-                raw_frames.is_empty(),
-                "when present HardFault handler must be the first frame we unwind but wasn't"
-            );
-
-            outcome = if overflowed_stack(sp, sp_ram_region) {
-                Outcome::StackOverflow
-            } else {
-                Outcome::HardFault
-            };
-        }
+        output.outcome = check_hard_fault(pc, vector_table, &mut output, sp, sp_ram_region)
+            .unwrap_or(output.outcome);
```

Comment:
The `.unwrap_or(output.outcome)` confused me at first, but it just mean that we just leave the value untouched if we return `None`?

Maybe if-let would be more explicit?

Reply:
duh! good point.done :D